rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+With that in mind, our goal for all of the proposed improvements will be to maximize the effectiveness of teams working with Terraform while still maintaining a secure environment. This means that any Terraform changes that include IAM permissions will still require infrastructure approval. It will also likely mean excluding certain IAM permissions from Scalr and/or engineering user groups, possibly including those that allow deletion of certain types of resources.
+
+For more details on how we'll accomplish this, see the corresponding [reference-level explanation][reference-permissions].
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how
+the detailed proposal makes those examples work.
+-->
+
+## 1. Splitting Root Modules By Team
+[reference-module-splitting]: #1-splitting-root-modules-by-team
+
+To optimize the management of development and staging resources, the `dev` root module will be split into multiple root modules that are organized by team. Once a new root module is set up for each engineering team, we'll need to:
+
+- Migrate each team's development and staging resources out of the `dev` module and into their new team-based module
+- Update references to dependent resources that remain in other modules to use data sources or [remote state](https://developer.hashicorp.com/terraform/language/state/remote-state-data)
+
+Note that production resources will remain in a single `prod` root module in the short term, but will likely eventually also be moved to team-based modules a later date. This will allow us to vet the changes proposed in this document, while also making sure we're adhering to the stricter security requirements for production that are discussed in the [Scalr workflow for production][scalr-workflow-production] section.
+
+### Root Module Naming Convention
+[root-module-naming-convention]: #root-module-naming-convention
+
+We'll also need to implement a new naming convention for root modules to keep them organized.
+
+Terraform's official [naming convention for published modules][terraform-module-publishing] is `terraform-<PROVIDER>-<NAME>`, where `<NAME>` can contain additional hyphens. Since our own root modules won't be published to Terraform's public registry, we can omit the `terraform-` prefix and use `<PROVIDER>-<NAME>` as a base convention, where `<NAME>` can contain any of the following:
```

Comment:
[RFC 13](https://github.com/mbta/technology-docs/blob/main/rfcs/accepted/0013-consistent-resource-names.md) specifies that Terraform names should be `snake_case`. If modules are an exception, they should probably be documented there as an exception.

Reply:
That's true of resource names but not module names. I'll make a task to update that document to clarify the distinction.